core/app/c/[communitySlug]/settings/tokens/PermissionField.tsx

```diff
@@ -11,13 +11,63 @@ import { Popover, PopoverContent, PopoverTrigger } from "ui/popover";
 
 import type { CreateTokenForm } from "./CreateTokenForm";
 
+/**
+ * This is a type for a configuration object for form fields. It allows you to specify
+ * what a certain Permission form field does in a type-safe way, depending on the earlier
+ * configuration of the `permissionsSchema`.
+ *
+ * Roughly this type means:
+ *
+ * Turn a type that looks like this,
+ * where `boolean | undefined` means that the field can be set to `true` or `false`, and
+ * `boolean | { stages: Stages[] } | undefined` means that the field can be set to `true` or `false`,
+ * or to an object with a `stages` field that is an array of stages:
+ * ```ts
+ * type InputShape =  {
+ * 	 community: {
+ * 		read: boolean | undefined
+ *   },
+ *   pub: {
+ * 		write: boolean | {
+ * 			stages: Stages[]
+ * 		} | undefined
+ *   }
+ * }
+ * ```
+ * into a type that looks like
+ * ```ts
+ * {
+ * 	 community: {
+ * 		read: ({ context, form, value, onChange }: {
+ * 			context: CreateTokenFormContext;
+ * 			form: CreateTokenForm;
+ * 			value: boolean;
+ * 			onChange: (...args: any[]) => void;
+ *        }) => JSX.Element
+ * },
+ *   pub: {
+ * 		write?: ({ context, form, value, onChange }: {
+ * 			context: CreateTokenFormContext;
+ * 			form: CreateTokenForm;
+ * 			value: boolean | {
+ * 				stages: Stages[]
+ * 			};
+ * 			onChange: (...args: any[]) => void;
+ *        }) => JSX.Element
+ *   } | undefined
+ * }
+ * ```
+ * The idea is that you can customize what form fields look like depending on the
+ * `permissionsSchema` configuration.
+ * E.g. the `write` field of the `pub` scope can be configured to only write to specific stages.
+ * You should be able to configure a custom element to do this.
+ *
+ */
```

Comment:
even i had some problems remembering what the hell this did, so i left a comment for future reference